compiler/src/dotty/tools/dotc/transform/LinkAll.scala

```diff
@@ -0,0 +1,78 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.{CompilationUnit, FromTasty}
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.SymDenotations.ClassDenotation
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.transform.TreeTransforms._
+
+/** Loads all potentially reachable trees from tasty. ▲
+ *  Only performed on whole world optimization mode. ▲ ▲
+ *
+ *  TODO: Next step is to only load compilation units reachable in the call graph
+ */
+class LinkAll extends MiniPhaseTransform {
+  import tpd._
+  import LinkAll._
+
+  override def phaseName = "linkAll"
+
+  override def prepareForUnit(tree: tpd.Tree)(implicit ctx: Context): TreeTransform = NoTransform
+
+  override def runOn(units: List[CompilationUnit])(implicit ctx: Context): List[CompilationUnit] =
+    if (ctx.settings.XlinkOptimise.value) super.runOn(allUnits(Set.empty, units.toSet, Set.empty))
+    else super.runOn(units)
+
+  /** Loads and processes new compilation units, possibly loading more units. */
+  private def allUnits(processed: Set[CompilationUnit], unprocessed: Set[CompilationUnit], loadedClasses: Set[ClassDenotation])(implicit ctx: Context): List[CompilationUnit] = {
+    if (unprocessed.isEmpty) processed.toList
+    else {
+      val accum = new ClassesToLoadAccumulator
+      val classesToLoad = unprocessed.foldLeft(Set.empty[ClassDenotation])((acc, unit) => accum.apply(acc, unit.tpdTree)) -- loadedClasses
+      val loadedUnits = classesToLoad.flatMap(cls => loadCompilationUnit(cls))
+      allUnits(processed ++ unprocessed, loadedUnits, loadedClasses ++ classesToLoad)
+    }
+  }
+
+  /** Collects all class denotations that may need to be loaded. */
+  private class ClassesToLoadAccumulator extends TreeAccumulator[Set[ClassDenotation]] {
+    private var inParents = false
+    override def apply(acc: Set[ClassDenotation], tree: tpd.Tree)(implicit ctx: Context): Set[ClassDenotation] = tree match {
+      case New(tpt) => accum(acc, tpt.tpe.classSymbol)
+      case AppliedTypeTree(tpt, _) if inParents => accum(acc, tpt.symbol)
+      case tree: RefTree if inParents || tree.symbol.is(Module) =>
+        foldOver(accum(acc, tree.symbol), tree)
+      case tree @ Template(constr, parents, self, _) =>
+        val acc1 = this(acc, constr)
+        inParents = true
+        val acc2 = this(acc1, parents)
+        inParents = false
```

Comment:
Don't think this is correct in general, you assume that inParents will be false before you set it to true, but you can have a Template inside the parents of a Template, e.g:
```scala
class A(x: Int)
class B extends A({ class Foo; 1 })
```

Reply:
This works, I added a test for it.